Replace pcl_isnan, pcl_isfinite and pcl_isinf by C++11 methods

SunBlack opened the pull request.

SunBlack commented:
Compile issue is raised here:

https://github.com/PointCloudLibrary/pcl/blob/99a5b28782b6751eb3c83a335806ae936080bf04/common/include/pcl/common/point_tests.h#L98-L102

Reason: `p.x` and `p.y` are of type int, but MS failed to implement `bool isfinite( IntegralType arg );` correct.

I could simply change code to:
```
 template <> inline bool 
 isFinite<pcl::BorderDescription> (const pcl::BorderDescription &p) 
 { 
   return (std::isfinite (static_cast<double>(p.x)) && std::isfinite (static_cast<double>(p.y))); 
 } 
```
But this should always return true, because it is and int type and can't store and infinite value, or? So maybe this is better fix:
```
 template <> inline bool 
 isFinite<pcl::BorderDescription> (const pcl::BorderDescription &p) 
 { 
   return true; 
 } 
```
What do you think?

taketwo commented:
Sure, integer can not be infinite; that test makes no sense. Just return `true`, perhaps move the specialization so that it's a part of the group of "always true" tests.

SergioRAgostinho commented on io/include/pcl/io/file_io.h:
```diff
@@ -246,37 +247,18 @@ namespace pcl
-      stream << "nan";
-    else
-      // Numeric cast doesn't give us what we want for int8_t
-      stream << boost::numeric_cast<int>(value);
```
I'm a little concerned over the removal of this specialization. When a value is of `int8_t`, printing to streams usually would dump a `char`, which is not what is desired here. Example:
```cpp
int8_t var = 0;
std::cout << var << std::endl;
```
This will either print a funky character or nothing at all, when when a type is `int8_t`, the user is really interested in seing a numerical value. Hence the cast and this particular specialization. By removing it entirely, we're gonna have the same problem again.

taketwo replied:
Good point, I agree that we should keep the byte specializations.

SunBlack replied:
Ah well, now I anderstand the comment `Numeric cast doesn't give us what we want for int8_t`. I readded these functions and changed comment, so next know why this type cast is necessary there.

But general `boost::numeric_cast` in general integral and floating point specialization is senseless, or? As far as I unterstand documentation this cast only makes sense in case types before and after cast are different.

SunBlack replied:
I removed unnecessary `boost::numeric_cast`. In case you see any reason for keeping this cast, tell me.

SergioRAgostinho commented on io/include/pcl/io/impl/pcd_io.hpp:
```diff
@@ -494,60 +494,42 @@ pcl::PCDWriter::writeASCII (const std::string &file_name, const pcl::PointCloud<
-              stream << "nan";
-            else
-              stream << boost::numeric_cast<uint32_t>(value);
+            stream << boost::numeric_cast<uint32_t>(value);
```
I think there's a bug here, not introduced by you. The cast should be to an a signed `int` type.

SunBlack replied:
`uint32_t` is an `unsigned int`, and Values of a `uint8_t` fits always into a `uint32_t`. So where is the bug?

taketwo replied:
We are in `int8` branch here (signed).

SunBlack replied:
Ups, was 5 lines below. Will fix it.

SergioRAgostinho commented on io/include/pcl/io/impl/pcd_io.hpp:
```diff
@@ -803,60 +782,42 @@ pcl::PCDWriter::writeASCII (const std::string &file_name,
-              stream << "nan";
-            else
-              stream << boost::numeric_cast<uint32_t>(value);
+            stream << boost::numeric_cast<uint32_t>(value);
```
Same issue here.

taketwo commented on io/include/pcl/io/file_io.h:
```diff
@@ -298,7 +280,8 @@ namespace pcl
     Type value;
     memcpy (&value, &cloud.data[point_index * point_size + cloud.fields[field_idx].offset + fields_count * sizeof (Type)], sizeof (Type));
-    if (!std::isfinite (value))
+    //MSVC is missing bool std::isfinite(IntegralType arg); variant, so we need to cast ourself to double
```
Can we create two versions just like with `copyValueString`? The integral version won't even need to `memcpy`, just return `true`.

taketwo commented on io/src/ply_io.cpp:
```diff
@@ -267,7 +267,8 @@ namespace pcl
   {
-    if (!std::isfinite (value))
+    //MSVC is missing bool std::isfinite(IntegralType arg); variant, so we need to cast ourself to double
+    if (!std::isfinite (static_cast<double>(value)))
```
What about if we add a local helper function, something like

```cpp
template<typename T> inline void
unsetDenseFlagIfNotFinite(T value, PCLPointCloud2& cloud)
{
  if (!std::isfinite)
    cloud.is_dense = false;
}
```

with a specialization for integral types that is a no-op?

SunBlack replied:
10 lines of code vs a single cast, but okay, I used template specialization for it ;).

taketwo replied:
Sorry for the trouble :)

SunBlack commented:
> The last commit has too many unrelated changes in my opinion. Could you please split it while addressing comments?

They are not unrelated. This are all MSVC fixes (Azure failed before these changes).

taketwo commented on io/src/ply_io.cpp:
```diff
@@ -294,7 +308,8 @@ namespace pcl
   PLYReader::vertexListPropertyContentCallback (ContentType value)
   {
-    if (!std::isfinite (value))
+    //MSVC is missing bool std::isfinite(IntegralType arg); variant, so we need to cast ourself to double
```
Use `unsetDenseFlagIfNotFinite(value, cloud_)` here as well?
